### PR TITLE
Welcome to version 2.0

### PR DESCRIPTION
**Pull request by karllindmark**
Battle Chat is being bumped to version 2.0

Changes include, but are not limited to:
- Gradle support!!!
- Navigation Drawer (huzzah!)
- Fragments
- Loaders
- Better looking UI (thanks @humphreybc!)
- Gson!

### CODE REVIEW

**Comment by karllindmark** [comment]
@peter-budo

**Comment by karllindmark** [comment]
@gustavkarlsson

File: app/src/main/java/com/ninetwozero/battlechat/BattleChat.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import android.app.Application;
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+
+import com.ninetwozero.battlechat.misc.Keys;
+
+public class BattleChat extends Application {
+    public final static String TAG = "com.ninetwozero.battlechat";
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/BattleChat.java]: Obsolete `getSimpleName()` is better

File: app/src/main/java/com/ninetwozero/battlechat/base/asynctasks/BaseLogoutTask.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+public class BaseLogoutTask extends AsyncTask<Void, Void, Void> {
+    private final Context context;
+
+    public BaseLogoutTask(final Context context) {
+        this.context = context;
+    }
+
+    @Override
+    protected Void doInBackground(Void... params) {
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/base/asynctasks/BaseLogoutTask.java]: params not used

File: app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListAdapter.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        return items;
+    }
+
+    final public void setItems(List<T> items) {
+        this.items = items;
+        notifyDataSetChanged();
+    }
+
+    final public void setText(final View container, final int resourceId, final String text) {
+        setText(container, resourceId, text, -1);
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListAdapter.java]: `-1` magic number

File: app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListFragment.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+public abstract class BaseListFragment extends ListFragment {
+    private Toast toast;
+    protected LayoutInflater layoutInflater;
+
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        super.onCreateView(inflater, container, savedInstanceState);
+        this.layoutInflater = inflater;
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListFragment.java:22]: Is that assignment necessary? Also returning null?

File: app/src/main/java/com/ninetwozero/battlechat/network/IntelLoader.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import android.content.Context;
+import android.support.v4.content.AsyncTaskLoader;
+
+import com.github.kevinsawicki.http.HttpRequest;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.ninetwozero.battlechat.factories.GsonFactory;
+import com.ninetwozero.battlechat.network.exception.Failure;
+
+public class IntelLoader<T> extends AsyncTaskLoader<Result> {
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/network/IntelLoader.java]: Thief! not even changing class name ROFL

File: app/src/main/java/com/ninetwozero/battlechat/datatypes/NoResultExpected.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * 	BattleChat is distributed in the hope that it will be useful,
+ * 	but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * 	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * 	GNU General Public License for more details.
+ * /
+ */
+
+package com.ninetwozero.battlechat.datatypes;
+
+public class NoResultExpected {
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/datatypes/NoResultExpected.java]: Weird class name. Am I suppose to understand that loader should not return any result and throw exception if there is some result?

File: app/src/main/java/com/ninetwozero/battlechat/datatypes/Session.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+*/
+
+package com.ninetwozero.battlechat.datatypes;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.os.Bundle;
+import android.preference.PreferenceManager;
+
+import com.ninetwozero.battlechat.misc.Keys;
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/datatypes/Session.java]: Cry of despair! `misc`

File: app/src/main/java/com/ninetwozero/battlechat/datatypes/Session.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+    public static void clearSession(final Context c) {
+        Session.sessionData.clear();
+
+        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(c);
+        SharedPreferences.Editor editor = preferences.edit();
+
+        editor.remove(Keys.Session.USER_ID);
+        editor.remove(Keys.Session.USERNAME);
+        // editor.remove(Keys.Session.EMAIL);
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/datatypes/Session.java]: Remove any commented out code

File: app/src/main/java/com/ninetwozero/battlechat/factories/UrlFactory.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    public static URL buildLogoutUrl() {
+        return createURL("session/logout/");
+    }
+
+    public static URL buildFriendListURL() {
+        return createURL("comcenter/initComcenter/");
+    }
+
+    public static URL buildOpenChatURL(final String userId) {
+        return createURL(String.format("comcenter/getChatId/%s/", userId));
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/factories/UrlFactory.java:38]: I realize we are not doing it in other project too. Any stream or string manipulation should include encoding `UTF-8`

File: app/src/main/java/com/ninetwozero/battlechat/interfaces/AboutInfoRow.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    private int title;
+    private int subTitle;
+    private String url;
+    private Type type;
+
+    public AboutInfoRow(final int title) {
+        this.title = title;
+        this.type = Type.HEADER;
+    }
+
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/interfaces/AboutInfoRow.java]: Extra line :tongue:

File: app/src/main/java/com/ninetwozero/battlechat/network/Result.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    public String getMessage() {
+        return message;
+    }
+
+    public T getData() {
+        return data;
+    }
+
+    @Override
+    public String toString() {
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/network/Result.java]: Unnecessary polution

File: app/src/main/java/com/ninetwozero/battlechat/services/BattleChatService.java [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+        @Override
+        protected Integer doInBackground(final Void... params) {
+            try {
+                Log.i(TAG, "Talking to the website...");
+                final SimpleGetRequest request = new SimpleGetRequest(UrlFactory.buildFriendListURL());
+                final ComCenterRequest comCenter = createComCenterRequestFromJson(request.execute());
+                final List<Chat> unreadChats = findUnreadChats(comCenter.getInformation().getChats());
+
+                Log.d(TAG, "Number of unread chats: " + unreadChats.size());
+                for (Chat chat : unreadChats) {
```
Comment (peter-budo) [inline app/src/main/java/com/ninetwozero/battlechat/services/BattleChatService.java]: Ugly for loop inside for loop

File: app/src/main/java/com/ninetwozero/battlechat/base/asynctasks/BaseLogoutTask.java (hunk 2 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/base/asynctasks/BaseLogoutTask.java]: Params not used, but has to be there due to how `doInBackground(...)` is being defined. `Void...` corresponds to blank parameters, in this case.

File: app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListFragment.java (hunk 4 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListFragment.java:22]: Yeah, it's to keep the instance for further use. :-)

Returning null won't be an issue in this case, as we don't use the return value. However, `onCreateView(LayoutInflater, ViewGroup, Bundle)` requires a view to be returned, but that's handled by the "kids".

File: app/src/main/java/com/ninetwozero/battlechat/network/IntelLoader.java (hunk 5 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/network/IntelLoader.java]: I'm quite transparent with what I've been copying between the projects - I'd like to see some sort of "Common Battlelog library" in the future if possible. For things like this, for example. :-D

File: app/src/main/java/com/ninetwozero/battlechat/datatypes/NoResultExpected.java (hunk 6 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/datatypes/NoResultExpected.java]: It's supposed to be some sort of "fire & forget", but the loaders themselves want us to pass a "datatype" for the generics. I suppose I could use `Void` instead, but not sure if it's better (from a readability point of view).

File: app/src/main/java/com/ninetwozero/battlechat/datatypes/Session.java (hunk 7 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/datatypes/Session.java]: Moving them to root package ;-)

File: app/src/main/java/com/ninetwozero/battlechat/factories/UrlFactory.java (hunk 9 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/factories/UrlFactory.java:38]: Do you have an example on how this usage would be? :cake:

File: app/src/main/java/com/ninetwozero/battlechat/services/BattleChatService.java (hunk 12 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/services/BattleChatService.java]: Fixed!

File: app/src/main/java/com/ninetwozero/battlechat/interfaces/AboutInfoRow.java (hunk 10 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/interfaces/AboutInfoRow.java]: Fixed!

File: app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListFragment.java (hunk 4 above)
Comment (karllindmark) [inline app/src/main/java/com/ninetwozero/battlechat/base/ui/BaseListFragment.java:22]: `Return the View for the fragment's UI, or null.` (from the docs for android.support.v4.app.ListFragment)

**Comment by karllindmark** [comment]
@peter-budo: Could you try this on your Nexus 7 and post a screenshot of 1) the login screen and 2) the "main" activity? That would be helpful for our future adventures. :+1:

**Pull request closed without merging**